pkg/deploymanager/subscription.go

```diff
@@ -0,0 +1,322 @@
+	operatorGroups []v1.OperatorGroup
+	catalogSources []v1alpha1.CatalogSource
+	subscriptions  []v1alpha1.Subscription
+}
+
+// DeployODFWithOLM deploys odf operator via an olm subscription
+func (t *DeployManager) DeployODFWithOLM(odfCatalogImage string, subscriptionChannel string) error {
+
+	if odfCatalogImage == "" {
+		return fmt.Errorf("catalog registry images not supplied")
+	}
+
+	co := t.generateClusterObjects(odfCatalogImage, subscriptionChannel)
+	err := t.deployClusterObjects(co)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (t *DeployManager) generateClusterObjects(odfCatalogImage string, subscriptionChannel string) *clusterObjects {
+
+	co := &clusterObjects{}
+	label := make(map[string]string)
+	// Label required for monitoring this namespace
+	label["openshift.io/cluster-monitoring"] = "true"
+
+	// Namespaces
+	co.namespaces = append(co.namespaces, k8sv1.Namespace{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Namespace",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   InstallNamespace,
+			Labels: label,
+		},
+	})
+
+	// Operator Groups
+	odfOG := v1.OperatorGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "openshift-storage-operatorgroup",
+			Namespace: InstallNamespace,
+		},
+		Spec: v1.OperatorGroupSpec{
+			TargetNamespaces: []string{InstallNamespace},
+		},
+	}
+
+	co.operatorGroups = append(co.operatorGroups, odfOG)
+
+	// Catalog Source
+	odfCatalog := v1alpha1.CatalogSource{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "odf-catalogsource",
+			Namespace: marketplaceNamespace,
+		},
+		Spec: v1alpha1.CatalogSourceSpec{
+			SourceType:  v1alpha1.SourceTypeGrpc,
+			Image:       odfCatalogImage,
+			DisplayName: "OpenShift Data Foundation",
+			Publisher:   "Red Hat",
+		},
+	}
+
+	co.catalogSources = append(co.catalogSources, odfCatalog)
+
+	odfCatalog = v1alpha1.CatalogSource{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "odf-catalogsource",
+			Namespace: InstallNamespace,
+		},
+		Spec: v1alpha1.CatalogSourceSpec{
+			SourceType:  v1alpha1.SourceTypeGrpc,
+			Image:       odfCatalogImage,
+			DisplayName: "OpenShift Data Foundation",
+			Publisher:   "Red Hat",
+		},
+	}
+
+	co.catalogSources = append(co.catalogSources, odfCatalog)
+
+	// Subscriptions
+	odfSubscription := v1alpha1.Subscription{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "odf-subscription",
+			Namespace: InstallNamespace,
+		},
+		Spec: &v1alpha1.SubscriptionSpec{
+			Channel:                subscriptionChannel,
+			InstallPlanApproval:    "Automatic",
+			Package:                "odf-operator",
+			CatalogSource:          "odf-catalogsource",
+			CatalogSourceNamespace: marketplaceNamespace,
+		},
+	}
+
+	co.subscriptions = append(co.subscriptions, odfSubscription)
+
+	return co
+}
+
+// Deploying the cluster objects
+func (t *DeployManager) deployClusterObjects(co *clusterObjects) error {
+
+	for _, namespace := range co.namespaces {
+		err := t.CreateNamespace(namespace.Name)
```

Comment:
As you are just using a name, why not to create a list of namespaces instead of creating a object.

Reply:
I think the clusterobject here has the namespace, operatorgroup, catalogsource & subscription. So it can not be a list I think

Reply:
It can be, lets discuss this over a call